Adds container data functionality and gets preferred container

p-galligan opened the pull request:
Adds functionality to get a string of all containers for an object, and gets the preferred container for that object. Addresses #20 and #21

Commit:
add get_container_indicators

Adds logic to get a list of all container types + indicators.

Adds items to the cassette list.

Commit:
update cassette

Adds a cassette for archival object with no instances

Commit:
move instance assignation out of method

Removes the assignation of instances out of get_container_indicators

Commit:
remove instance check

Removes the instance check from `get_containers` and leaves that logic in `get_data` as we'll be checking for preferred formats there at the same time.

Updates adds `get_containers` tests.

Commit:
futzing with logic

necessary instance and container information retrieved, just figuring out how I want to interact with it

Commit:
get preferred containers

code in place to get preferred containers and also container-specific fields for as_data

Commit:
update set_preferred_data

Updates how we set preferred data and adds a few tests.

Commit:
update test routines

remove variable assignation in testroutines

p-galligan committed:
Merge branch 'development' into issue-20

Commit:
fix linting

linting fixes

Commit:
Merge branch 'issue-20' of https://github.com/RockefellerArchiveCenter/request_broker into issue-20

p-galligan commented:
Still a WIP (don't have tests for `get_instance_data` or `get_preferred_format` yet, which I'd take some advice on for those given that they're a bit trickier than I'm used to.

Otherwise, I wouldn't mind a second set of eyes at this point to see if the rest makes sense, particularly lines 55-71 in `routines.py` and the corresponding methods.

I've separated out `set_preferred_data` and called it three times because I figured that we'd want those fields even if there are no containers, or a preferred item. However, for no containers, I could see skipping them since we wouldn't ever use them.

helrond commented on process_request/helpers.py:
```diff
@@ -3,6 +3,164 @@
+        string: A concatenated string containing the container type and container indicator, or digital object title.
+    """
+    if instance.get("instance_type") == "digital_object":
+        return "Digital Object: " + instance.get("digital_object").get("_resolved").get("title")
```
```suggestion
        return "Digital Object:  {}".format(instance.get("digital_object").get("_resolved").get("title"))
```
Format strings are easier to read!

helrond commented on process_request/helpers.py:
```diff
@@ -3,6 +3,164 @@
+    if instance.get("instance_type") == "digital_object":
+        return "Digital Object: " + instance.get("digital_object").get("_resolved").get("title")
+    else:
+        return instance.get("sub_container").get("top_container").get("_resolved").get("type").capitalize() + ' ' + instance.get("sub_container").get("top_container").get("_resolved").get("indicator")
```
Trying to make this a bit more readable. 

```suggestion
        top_container = instance.get("sub_container").get("top_container").get("_resolved")
        return "{} {}".format(top_container.get("type").capitalize(), top_container.get("indicator"))
```

helrond commented on process_request/helpers.py:
```diff
@@ -3,6 +3,164 @@
+    Returns:
+        string: all file version uris associated with the digital object, separated by a comma.
+    """
+    versions = []
+    [versions.append(v.get("file_uri")) for v in digital_object.get("file_versions")]
```
I think you were trying to do this:

```suggestion
    versions = [v.get("file_uri")) for v in digital_object.get("file_versions")]
```

p-galligan replied:
I forgot that you could assign to a list with list comprehension. So yeah, this is much simpler, thanks!

Commit:
change declaration to none

remove `''` and change it to None

helrond commented on process_request/helpers.py:
```diff
@@ -3,6 +3,164 @@
+    if type == "digital_object":
+        digital_object = instance.get("digital_object").get("_resolved")
+        instance_data["location"] = get_file_versions(digital_object)
+        instance_data["barcode"] = ''
```
Is there a reason we're not using `None`?

p-galligan replied:
I updated this while you were reviewing. No reason. It's now set to `None`

helrond commented on process_request/helpers.py:
```diff
@@ -3,6 +3,164 @@
+    Returns:
+        preferred (list): a list of string instance information retrieved by get_instance_data.
+    """
+    instance_types = []
+    [instance_types.append((i["instance_type"])) for i in instances]
```
Again, I think you meant:

```suggestion
    instance_types = [i["instance_type"] for i in instances]
```

helrond replied:
Although...I'm wondering why this. `instance_types` variable is even being set because it's not used...

p-galligan replied:
I'm actually using it in the if statements immediately below. `if "digital_object" in instances` wasn't working to return anything.

helrond commented on process_request/helpers.py:
```diff
@@ -3,6 +3,164 @@
+    [instance_types.append((i["instance_type"])) for i in instances]
+    if "digital_object" in instance_types:
+        preferred = [get_instance_data(i, i["instance_type"]) for i in instances if i["instance_type"] == "digital_object"]
+    elif "microform" in instance_types:
+        preferred = [get_instance_data(i, i["instance_type"]) for i in instances if i["instance_type"] == "microform"]
+    elif "microfilm" in instance_types:
+        preferred = [get_instance_data(i, i["instance_type"]) for i in instances if i["instance_type"] == "microfilm"]
```
Do we actually have both of these formats and if so, are they in the right order (i.e. would we want microform before microfilm)?

p-galligan replied:
For some reason I thought we have both. But a check of production AS reveals just microform, which is 👍

helrond commented on process_request/helpers.py:
```diff
@@ -3,6 +3,164 @@
+        preferred = [get_instance_data(i, i["instance_type"]) for i in instances if i["instance_type"] == "microform"]
+    elif "microfilm" in instance_types:
+        preferred = [get_instance_data(i, i["instance_type"]) for i in instances if i["instance_type"] == "microfilm"]
+    elif "mixed materials" in instance_types:
+        preferred = [get_instance_data(i, i["instance_type"]) for i in instances if i["instance_type"] == "mixed materials"]
+    else:
+        return None
```
I think if we've established there are no digital objects or microfilm instances, we can just return whatever else is there.

p-galligan replied:
So, we should still return that data even if it is not in one of our approved "delivery formats"?

helrond commented on process_request/helpers.py:
```diff
@@ -3,6 +3,164 @@
+    return preferred
+
+
+def set_preferred_data(data, indicator=None, type=None, location=None):
+    """Takes a dictionary and sets arguments to key, value pairs in that dictionary.
+
+    Args:
+        data (dict): the dictionary to write to.
+        indicator (string): string representation of a container type and indicator
+        type (string): string representation of an instance type
+        location (string): string representation of a container's location
+
+    Return:
+        data (dict): returns an updated dictionary with new key, value pairs
+    """
+    data['preferred_container'] = indicator
+    data['preferred_format'] = type
+    data['preferred_location'] = location
+    return data
```
I'm not sure I understand the utility of this, but that also may be tied to my confusion about what's happening in the main `get_data` method.

p-galligan replied:
See comment below. I originally didn't have this as a separate method, and I'm happy to remove it if we don't care about some of those key/value pairs showing up in the returned data.

helrond commented on process_request/tests.py:
```diff
@@ -55,6 +67,39 @@ def test_machineuser(self):
+class TestHelpers(TestCase):
+
+    def test_get_container_indicators(self):
+        letters = ''.join(random.sample(string.ascii_lowercase, random.randint(2, 10)))
```
You call this particular combination of things a number of times, so I'd suggest creating a `random_string()` helper.

helrond commented on process_request/tests.py:
```diff
@@ -55,6 +67,39 @@ def test_machineuser(self):
+
+    def test_get_container_indicators(self):
+        letters = ''.join(random.sample(string.ascii_lowercase, random.randint(2, 10)))
+        title = "Digital Object: " + letters
```
I would call this `expected_title` or something more descriptive. And you should use format strings rather than concatenation like this.

helrond commented on process_request/routines.py:
```diff
@@ -29,36 +32,46 @@ def get_data(self, item):
+            as_data['title'] = item_json.get("display_string")
+            as_data['ref'] = item_json.get("uri")
+            instances = item_json.get("instances")
+            if instances:
+                container_indicators = []
+                for i in instances:
+                    [container_indicators.append(get_container_indicators(i))]
+                as_data['containers'] = ", ".join(container_indicators)
+                preferred_item = get_preferred_format(instances)
+                if preferred_item:
+                    set_preferred_data(as_data,
+                                       preferred_item[0]['indicator'],
+                                       preferred_item[0]['instance_type'],
+                                       preferred_item[0]['location'])
+                else:
+                    set_preferred_data(as_data)
+            else:
+                as_data['containers'] = None
+                set_preferred_data(as_data)
```
This logic is super complex and I'm not sure I understand why?

p-galligan replied:
So, I think most of the complexity comes from the assumption that if I'm working with an `as_data` dict, that we'd want certain keys to show up even if they're empty.

First, I'm getting a list of all indicators joined together.

Then, I'm using the instances info to grab the preferred format. It's separated out from the `for i in instances` loop because I'm getting a list of `instance_types` in that method. However, I could just as easily remove that from the method and have `get_preferred_format` to use a single instance instead.

Then, the ending is kinda complex because of the assumption that we'd want those keys. As it has been written, there will not be a `preferred_item` in the cases where an instance type was not digital object, microform, or mixed materials. So, it was going to error if you based it av materials or something else. We could remove that check if we're returning something no matter what. That would also eliminate the need for the else statement.

The else statement at the end also assumes we'd want those keys even if they're set to `None`. They weren't showing up in `as_data` unless specifically set.

helrond replied:
OK, so first I'd think about encapsulating all the logic you need in your helper methods so that you always get the desired value back. That way you can keep this `get_data` method fairly clean, and it will be easier to read and debug the code in the future. So, for example, your `get_container_indicators` method should just return None if there are no instances, same for `get_preferred_format`.

Regarding your assumptions, I think it's true that we'd want keys to exist even if their value is null. However, I think that we *do* always want to return preferred instance values. If there aren't digital or microfilm instances then it's just whatever else is there.

helrond commented on process_request/routines.py:
```diff
@@ -29,36 +32,46 @@ def get_data(self, item):
+            as_data['ref'] = item_json.get("uri")
+            instances = item_json.get("instances")
+            if instances:
+                container_indicators = []
+                for i in instances:
+                    [container_indicators.append(get_container_indicators(i))]
```
See my earlier comments on list comprehension.

helrond commented on process_request/tests.py:
```diff
@@ -55,6 +67,39 @@ def test_machineuser(self):
+        indicator = get_container_indicators(instance)
+        self.assertEqual(indicator, title)
+
+        type = ''.join(random.sample(string.ascii_lowercase, random.randint(2, 10))).capitalize()
```
The problem here is that you're not testing the `capitalize()` in `get_container_indicators()`.  I'd call `capitalize()` below in line 82 instead.

helrond commented on process_request/tests.py:
```diff
@@ -55,6 +67,39 @@ def test_machineuser(self):
+        type = ''.join(random.sample(string.ascii_lowercase, random.randint(2, 10))).capitalize()
+        number = ''.join(random.sample(string.ascii_lowercase, random.randint(2, 10)))
+        instance = {'instance_type': 'mixed materials', 'sub_container': {'top_container': {'_resolved': {'type': type, 'indicator': number}}}}
+        combined = type + ' ' + number
```
Use format strings!

helrond reviewed (changes requested):
This is coming together nicely! A couple of things to think about, including a couple of places where list comprehension wasn't quite right. May be useful to review this tutorial: https://www.programiz.com/python-programming/list-comprehension

p-galligan committed:
Update process_request/helpers.py

changes to format string in getting digital object indicator

Co-authored-by: Hillel Arnold <helrond@hotmail.com>

p-galligan committed:
Update process_request/helpers.py

changes formatting in getting container indicators to use format strings

Co-authored-by: Hillel Arnold <helrond@hotmail.com>

p-galligan committed:
Update process_request/helpers.py

correct list comprehension in get file versions

Co-authored-by: Hillel Arnold <helrond@hotmail.com>

p-galligan committed:
Update process_request/helpers.py

fix list comprehension

Co-authored-by: Hillel Arnold <helrond@hotmail.com>

Commit:
update tests, add test helpers

Updates tests and adds `test_helpers.py` to make random strings.

Fixes some list comprehension and adds format strings.

Commit:
reworks some helpers and adds tests

Adds tests for each helper and reworks them to always return an expected type. Adds/changes some fixtures.

p-galligan committed:
Merge branch 'development' into issue-20

Commit:
update as_data test

Updates test_as_data to include resolved digital objects and correct return length

helrond committed:
use tuple expansion for readability

helrond committed:
remove unused method

helrond committed:
cleaning up docstrings and general tidying up

helrond committed:
refactor get_preferred_format and get_instance_data to clarify logic

helrond committed:
add additional todos, call set

helrond commented:
@p-galligan OK, I tweaked a bunch of things here:
- passing an instance list (rather than a single instance) to `get_instance_data` which avoids some of the complexity that needed to be introduced in `get_formats`.
- Using tuple expansion rather than indexing for readability.

There's still a bit of work to be done, as noted in the TODO for `get_preferred_formats`.

helrond committed:
add better parsing for instance values

helrond committed:
make use of json_from_fixture helper

helrond reviewed (approved):
I think this is mergeable. The one thing I'm not sure about is whether the test fixtures cover all the relevant scenarios for instances, but that's a think we can build out more later.

p-galligan commented:
@helrond Ok, we can talk about what more coverage we'd need for those tests/fixtures for instances.

I actually have a question about `prepare_values`. I was going to set up a test for that method, but it looks like that isn't preserving the original order of each list. Should we be using an OrderedSet?

helrond commented:
> I actually have a question about `prepare_values`. I was going to set up a test for that method, but it looks like that isn't preserving the original order of each list. Should we be using an OrderedSet?

It sounds like you know the answer to that question :)

Commit:
Adds OrderedSet

Updates `prepare_values` to use imported OrderedSet. Adds test for `prepared_values`
